In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt #graphs and visualizations
import seaborn as sns #graphs and visualizations
# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/voting-data-csv/S04AC223.csv


# **Loading Data**
Let's have a look at the data.

In [2]:
voter_data_path = ('../input/voting-data-csv/S04AC223.csv')
voter_data = pd.read_csv(voter_data_path)
voter_data.head()

/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3441: DtypeWarning: Columns (5) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


AC_No    ACNameEn  PollingBoothNo  SectionNo  votersl houseno       VIDNo  \
0    223  Aurangabad               1          1        1       1  NCY2630614   
1    223  Aurangabad               1          1        2       2  NCY2464733   
2    223  Aurangabad               1          1        3       2  NCY2467843   
3    223  Aurangabad               1          1        4       2  NCY2467850   
4    223  Aurangabad               1          1        5       3  NCY0415406   

           VoterNameEng            Votername     Sex  relationNameEng  \
0          saurav kumar           सौरव कुमार    Male     ashok prasad   
1  Sanjeev kumar Pandey  संजीव कुमार पाण्डेय    Male    Suresh Pandey   
2        Durgawati Devi      दुर्गावती` देवी  Female  Dudhnath paswan   
3       Dudhnath Paswan        दुधनाथ पासवान    Male    chhedi Paswan   
4          Sanjit kumar          संजीत कुमार    Male  Ramanand Paswan   

      relationname relation  Age SectionName SectionNameEng  \
0      अशोक प्रसाद   Father   22      जम्होर         Jamhor   
1    सुरेश पाण्डेय   Father   27      जम्होर         Jamhor   
2    दुधनाथ पासवान  Husband   57      जम्होर         Jamhor   
3      छेदी पासवान   Father   68      जम्होर         Jamhor   
4  रामानन्द पासवान   Father   24      जम्होर         Jamhor   

  PollingStationAddressEn PollingStationAddress  
0  Uchcha Vighalay Jamhor   उच्च विघालय  जम्होर  
1  Uchcha Vighalay Jamhor   उच्च विघालय  जम्होर  
2  Uchcha Vighalay Jamhor   उच्च विघालय  जम्होर  
3  Uchcha Vighalay Jamhor   उच्च विघालय  जम्होर  
4  Uchcha Vighalay Jamhor   उच्च विघालय  जम्होर

Let me first remove the redundant columns.

In [3]:
eng_voter_data = voter_data.drop(['Votername','relationname','SectionName','PollingStationAddress'],axis=1)
eng_voter_data.head()

AC_No    ACNameEn  PollingBoothNo  SectionNo  votersl houseno       VIDNo  \
0    223  Aurangabad               1          1        1       1  NCY2630614   
1    223  Aurangabad               1          1        2       2  NCY2464733   
2    223  Aurangabad               1          1        3       2  NCY2467843   
3    223  Aurangabad               1          1        4       2  NCY2467850   
4    223  Aurangabad               1          1        5       3  NCY0415406   

           VoterNameEng     Sex  relationNameEng relation  Age SectionNameEng  \
0          saurav kumar    Male     ashok prasad   Father   22         Jamhor   
1  Sanjeev kumar Pandey    Male    Suresh Pandey   Father   27         Jamhor   
2        Durgawati Devi  Female  Dudhnath paswan  Husband   57         Jamhor   
3       Dudhnath Paswan    Male    chhedi Paswan   Father   68         Jamhor   
4          Sanjit kumar    Male  Ramanand Paswan   Father   24         Jamhor   

  PollingStationAddressEn  
0  Uchcha Vighalay Jamhor  
1  Uchcha Vighalay Jamhor  
2  Uchcha Vighalay Jamhor  
3  Uchcha Vighalay Jamhor  
4  Uchcha Vighalay Jamhor

Characterstics of our data

In [4]:
eng_voter_data.describe()

AC_No  PollingBoothNo     SectionNo       votersl           Age
count  93384.0    93384.000000  93384.000000  93384.000000  93384.000000
mean     223.0       49.954521      1.368082    498.480125     42.038219
std        0.0       28.860800      0.832521    315.790702     15.816858
min      223.0        1.000000      1.000000      1.000000     18.000000
25%      223.0       25.000000      1.000000    235.000000     29.000000
50%      223.0       50.000000      1.000000    472.000000     40.000000
75%      223.0       75.000000      1.000000    728.000000     52.000000
max      223.0      100.000000      7.000000   1440.000000    122.000000

# **Answering Questions about the Dataset**

Q1 : Which PollingBoothNo has the maximum number of under 30 female population?
First let us create the appropiate dataset.

In [5]:
young_female_data = eng_voter_data.loc[(eng_voter_data['Sex'] == 'Female')&(eng_voter_data['Age']<30)]
young_female_data.head()

AC_No    ACNameEn  PollingBoothNo  SectionNo  votersl houseno       VIDNo  \
27    223  Aurangabad               1          1       28       4  NCY1108620   
40    223  Aurangabad               1          1       41       6  NCY1788728   
51    223  Aurangabad               1          1       52       7  NCY2577922   
69    223  Aurangabad               1          1       70       9  NCY2467892   
79    223  Aurangabad               1          1       80      10  NCY1795202   

    VoterNameEng     Sex   relationNameEng relation  Age SectionNameEng  \
27    Switi Devi  Female      rajesh kumar  Husband   26         Jamhor   
40   Ritu Kumari  Female  Dukhdevan Paswan   Father   24         Jamhor   
51  PRITI KUMARI  Female       RAMA PRASAD   Father   20         Jamhor   
69   Poonam Devi  Female      Manish Kumar  Husband   25         Jamhor   
79   Urmila devi  Female       Sunil yadav  Husband   24         Jamhor   

   PollingStationAddressEn  
27  Uchcha Vighalay Jamhor  
40  Uchcha Vighalay Jamhor  
51  Uchcha Vighalay Jamhor  
69  Uchcha Vighalay Jamhor  
79  Uchcha Vighalay Jamhor

In [6]:
print(young_female_data.PollingBoothNo.value_counts())

53    204
48    188
54    173
68    170
98    169
     ... 
80     52
10     50
71     47
5      42
45     29
Name: PollingBoothNo, Length: 100, dtype: int64


**Ans 1 : Polling Booth No 53 has the maximum number of under 30 females, which is 204.**

Q2 : Share a list of Head of family (HoF) for each household. HoF are individuals with maximum age in each family. A household is defined as members with same houseno grouped under PollingBoothNo, SectionNo


In [7]:
HoF = eng_voter_data.groupby(['PollingBoothNo','SectionNo','houseno']).apply(lambda x: x.loc[x.Age.idxmax()])
HoF.head()

AC_No    ACNameEn  PollingBoothNo  \
PollingBoothNo SectionNo houseno                                      
1              1         1          223  Aurangabad               1   
                         2          223  Aurangabad               1   
                         3          223  Aurangabad               1   
                         4          223  Aurangabad               1   
                         5          223  Aurangabad               1   

                                  SectionNo  votersl houseno       VIDNo  \
PollingBoothNo SectionNo houseno                                           
1              1         1                1        1       1  NCY2630614   
                         2                1        4       2  NCY2467850   
                         3                1        6       3  FPX5573233   
                         4                1       18       4  NCY0507723   
                         5                1       34       5  NCY1080043   

                                     VoterNameEng     Sex   relationNameEng  \
PollingBoothNo SectionNo houseno                                              
1              1         1           saurav kumar    Male      ashok prasad   
                         2        Dudhnath Paswan    Male     chhedi Paswan   
                         3            Karami Devi  Female      Arjun Paswan   
                         4         Etavariya Devi  Female  Lakeshwar Paswan   
                         5         Krishna Paswan    Male    Mahadev Paswan   

                                 relation  Age SectionNameEng  \
PollingBoothNo SectionNo houseno                                
1              1         1         Father   22         Jamhor   
                         2         Father   68         Jamhor   
                         3        Husband   58         Jamhor   
                         4        Husband   78         Jamhor   
                         5         Father   61         Jamhor   

                                 PollingStationAddressEn  
PollingBoothNo SectionNo houseno                          
1              1         1        Uchcha Vighalay Jamhor  
                         2        Uchcha Vighalay Jamhor  
                         3        Uchcha Vighalay Jamhor  
                         4        Uchcha Vighalay Jamhor  
                         5        Uchcha Vighalay Jamhor

**Answer 2 : I will provide the csv file of the dataframe. A short preview has been shown above. ******

In [8]:
HoF.to_csv('HoF_data.csv',index=False)

Q3 : Split the VoterNameEng into FirstNameEng, MiddleNameEng and LastNameEng. 

In [9]:
splitted = eng_voter_data['VoterNameEng'].str.split()
eng_voter_data['FirstNameEng'] = splitted.str[0]
eng_voter_data['LastNameEng'] = splitted.str[-1]
middle = splitted.str[1] 
eng_voter_data['MiddleNameEng'] = middle.mask(middle == eng_voter_data['LastNameEng'], '')
eng_voter_data.head()

AC_No    ACNameEn  PollingBoothNo  SectionNo  votersl houseno       VIDNo  \
0    223  Aurangabad               1          1        1       1  NCY2630614   
1    223  Aurangabad               1          1        2       2  NCY2464733   
2    223  Aurangabad               1          1        3       2  NCY2467843   
3    223  Aurangabad               1          1        4       2  NCY2467850   
4    223  Aurangabad               1          1        5       3  NCY0415406   

           VoterNameEng     Sex  relationNameEng relation  Age SectionNameEng  \
0          saurav kumar    Male     ashok prasad   Father   22         Jamhor   
1  Sanjeev kumar Pandey    Male    Suresh Pandey   Father   27         Jamhor   
2        Durgawati Devi  Female  Dudhnath paswan  Husband   57         Jamhor   
3       Dudhnath Paswan    Male    chhedi Paswan   Father   68         Jamhor   
4          Sanjit kumar    Male  Ramanand Paswan   Father   24         Jamhor   

  PollingStationAddressEn FirstNameEng LastNameEng MiddleNameEng  
0  Uchcha Vighalay Jamhor       saurav       kumar                
1  Uchcha Vighalay Jamhor      Sanjeev      Pandey         kumar  
2  Uchcha Vighalay Jamhor    Durgawati        Devi                
3  Uchcha Vighalay Jamhor     Dudhnath      Paswan                
4  Uchcha Vighalay Jamhor       Sanjit       kumar

**Ans 3 : I will provide the csv file of the dataframe. A short preview has been shown above.**

In [10]:
splitted_data = eng_voter_data.iloc[:,[0,1,2,3,4,5,6,7,14,16,15,8,9,10,11,12,13]]
splitted_data.head()
splitted_data.to_csv('splitted_data.csv',index=False)

Q4 : What is the RelativeName of VoterNameEng ‘Gajala Praveen’ living in PollingBoothNo 19? 

In [11]:
relative = eng_voter_data.loc[(eng_voter_data['VoterNameEng'] == 'Gajala Praveen')&(eng_voter_data['PollingBoothNo']==19)]
relative.head()

AC_No    ACNameEn  PollingBoothNo  SectionNo  votersl houseno  \
17187    223  Aurangabad              19          1      171      45   

            VIDNo    VoterNameEng     Sex relationNameEng relation  Age  \
17187  NCY1155795  Gajala Praveen  Female    Sartaj Ahmad  Husband   27   

      SectionNameEng    PollingStationAddressEn FirstNameEng LastNameEng  \
17187   BHOJA BIGAHA  Prathamik Vidyalay Saheya       Gajala     Praveen   

      MiddleNameEng  
17187

**Ans 4 : The name of the relative is Sartaj Ahmad.**

Q5 : Prepare frequency distribution of the FirstName of VoterNameEng. What inference can you draw from this distribution?  

In [12]:
count = eng_voter_data['FirstNameEng'].value_counts()
print(count)

Ram                1476
Mohammad           1377
Sunita              552
Sanjay              534
Shanti              520
                   ... 
NandKishor            1
Shivdutt              1
Vindhyayavasini       1
rukhihi               1
Shaimal               1
Name: FirstNameEng, Length: 14703, dtype: int64


**Ans 5 :We can see that 'Ram' and 'Mohammad' are by far the most common first names in the database. These are names based on religious deities in Hinduism and Islam respectively. Both are also male names, which also gives a clue regarding the higher male population in the dataset.Also, there are a large amount of names with very low frequency of occurence. The top 1% of the most popular names account for almost 40% of the population. I will add a csv file of the same.**

In [13]:
count.to_csv('Frequency_counts_of_first_names.csv')

Q6 : Share the percentage distribution of each surname (last name) of VoterNameEng

In [14]:
last_name_counts = (eng_voter_data['LastNameEng'].value_counts())*(100/93384)
print(last_name_counts)

Devi              29.799537
Singh              7.624432
Kumar              5.719395
Yadav              4.735287
Ram                4.371198
                    ...    
n                  0.001071
pratap             0.001071
Chandharawansi     0.001071
partap             0.001071
Paramahans         0.001071
Name: LastNameEng, Length: 3044, dtype: float64


In [15]:
last_name_counts.to_csv('Percentage_Distribution_of_each_surname.csv')

**Ans 6 : 'Devi' is the most popular surname accounting for almost 30% of the population. Followed by 'Singh', which is a distant second. I will upload a csv file of the same.**

Q7 : How many families are there in the dataset? Group_by on PollingBoothNo, SectionNo, houseno

Since, the HoF.csv file already contains the head of family of each family, the number of rows in that dataframe will give us the number of families in the dataset.

In [16]:
print(len(HoF.index))

13025


**Ans 7 : There are 13025 famillies in the dataset.**

Q8 : As per dataset, what is the average family size in the households? 

In [17]:
family_data=pd.DataFrame()
family_data=eng_voter_data.groupby(['PollingBoothNo','SectionNo','houseno']).size()
#family_data.head()
family_data.reset_index()

PollingBoothNo  SectionNo houseno   0
0                   1          1       1   1
1                   1          1       2   3
2                   1          1       3   7
3                   1          1       4  23
4                   1          1       5   3
...               ...        ...     ...  ..
13020             100          3       5   5
13021             100          3       6  13
13022             100          3       7  15
13023             100          3       8  40
13024             100          3       9   5

[13025 rows x 4 columns]

In [18]:
family_data.to_csv('family_size.csv')

**Ans 8: From the above csv file, we can easily calculate the average family size, which comes out to be approximately 7.2 persons per household. Therefore, we can say that on an average there are 7 people in each family.**

Q9 : Which area has the highest population? An area is the same as SectionNameEng.

In [19]:
area_data=eng_voter_data.groupby('SectionNameEng').size()
area_data.reset_index()
area_df = pd.DataFrame(area_data,columns=['Population'])
print(area_df.idxmax())
print(area_df.max())
area_df.to_csv('Area_Population.csv')

Population    Jamhor
dtype: object
Population    5241
dtype: int64


**Ans 9: Therefore, we can see that Jamhor has the highest population, which is 5241.**

Q10 : As you see in the database, VIDNo is the voter id number of the voters. It is personal information and should not be shared publicly due to privacy concerns. How will you mask the middle 5 digits of the VIDNo? Mask 4th, 5th, 6th, 7th, 8th digit of the VIDNo. 

In [20]:
masked_ID = eng_voter_data['VIDNo']
masked_ID = masked_ID.astype(str)
def mask_voter_ID(x):    
    return (x[0:3] + '#####' + x[8:10])
masked_ID = masked_ID.apply(mask_voter_ID)
masked_ID.head()

0    NCY#####14
1    NCY#####33
2    NCY#####43
3    NCY#####50
4    NCY#####06
Name: VIDNo, dtype: object

In [21]:
masked_dataset = eng_voter_data.drop(['VIDNo'],axis=1)
masked_dataset['VIDNo'] = masked_ID
masked_dataset = masked_dataset.iloc[:,[0,1,2,3,4,5,16,6,7,8,9,10,11,12,13,14,15]]
masked_dataset.to_csv('Hidden_VIDNo.csv')
masked_dataset.head()

AC_No    ACNameEn  PollingBoothNo  SectionNo  votersl houseno       VIDNo  \
0    223  Aurangabad               1          1        1       1  NCY#####14   
1    223  Aurangabad               1          1        2       2  NCY#####33   
2    223  Aurangabad               1          1        3       2  NCY#####43   
3    223  Aurangabad               1          1        4       2  NCY#####50   
4    223  Aurangabad               1          1        5       3  NCY#####06   

           VoterNameEng     Sex  relationNameEng relation  Age SectionNameEng  \
0          saurav kumar    Male     ashok prasad   Father   22         Jamhor   
1  Sanjeev kumar Pandey    Male    Suresh Pandey   Father   27         Jamhor   
2        Durgawati Devi  Female  Dudhnath paswan  Husband   57         Jamhor   
3       Dudhnath Paswan    Male    chhedi Paswan   Father   68         Jamhor   
4          Sanjit kumar    Male  Ramanand Paswan   Father   24         Jamhor   

  PollingStationAddressEn FirstNameEng LastNameEng MiddleNameEng  
0  Uchcha Vighalay Jamhor       saurav       kumar                
1  Uchcha Vighalay Jamhor      Sanjeev      Pandey         kumar  
2  Uchcha Vighalay Jamhor    Durgawati        Devi                
3  Uchcha Vighalay Jamhor     Dudhnath      Paswan                
4  Uchcha Vighalay Jamhor       Sanjit       kumar

**Ans 10 : To mask the VIDNo, I have first converted the data to 'str' type, for easier manipulation. Then I have simply replaced the 4th,5th,6th,7th,8th digits with '#' and stored it in a new dataframe, so that I still have access to the original data. I will upload a csv file of the same.**